sbom/spdx/encoder.go

```diff
@@ -0,0 +1,444 @@
+	CreatorType string
+}
+
+var _ sbom.Encoder = (*Encoder)(nil)
+
+type Encoder struct {
+	Version           Version
+	Format            Format
+	Creators          []Creator
+	DocumentName      string
+	DocumentNamespace string
+	DocumentComment   string
+}
+
+func NewDefaultEncoder(options ...Option) *Encoder {
+	e := &Encoder{
+		Version: V2_3,
+		Format:  JSONFormat,
+		Creators: []Creator{
+			{
+				Creator:     "Claircore-" + getVersion(),
+				CreatorType: "Tool",
+			},
+		},
+	}
+
+	for _, opt := range options {
+		opt(e)
+	}
+
+	return e
+}
+
+func WithDocumentName(name string) Option {
+	return func(e *Encoder) {
+		e.DocumentName = name
+	}
+}
+
+func WithDocumentNamespace(namespace string) Option {
+	return func(e *Encoder) {
+		e.DocumentNamespace = namespace
+	}
+}
+
+func WithDocumentComment(comment string) Option {
+	return func(e *Encoder) {
+		e.DocumentComment = comment
+	}
+}
+
+// Encode encodes a [claircore.IndexReport] that writes to w.
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, w io.Writer, ir *claircore.IndexReport) error {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return err
+	}
+
+	// TODO(blugo): support SPDX versions before 2.3
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return fmt.Errorf("unknown SPDX version: %v", e.Version)
+	}
+
+	switch e.Format {
+	case JSONFormat:
+		if err := spdxjson.Write(tmpConverterDoc, w); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return fmt.Errorf("unknown requested format: %v", e.Format)
+}
+
+func (e *Encoder) parseIndexReport(ctx context.Context, ir *claircore.IndexReport) (*v2_3.Document, error) {
+	creators := make([]v2common.Creator, len(e.Creators))
+	for i, creator := range e.Creators {
+		creators[i] = v2common.Creator{
+			Creator:     creator.Creator,
+			CreatorType: creator.CreatorType,
+		}
+	}
+
+	// Initial metadata
+	out := &v2_3.Document{
+		SPDXVersion:       v2_3.Version,
+		DataLicense:       v2_3.DataLicense,
+		SPDXIdentifier:    "DOCUMENT",
+		DocumentName:      e.DocumentName,
+		DocumentNamespace: e.DocumentNamespace,
+		CreationInfo: &v2_3.CreationInfo{
+			Creators: creators,
+			Created:  time.Now().Format("2006-01-02T15:04:05Z"),
+		},
+		DocumentComment: e.DocumentComment,
+	}
+
+	pkgs := make(map[int]*v2_3.Package)
+	var pkgIds []int
+	dists := make(map[int]*v2_3.Package)
+	var distIds []int
+	repos := make(map[int]*v2_3.Package)
+	var repoIds []int
+	pkgRels := map[int][]*v2_3.Relationship{}
+	for _, r := range ir.IndexRecords() {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		rPkgId, err := strconv.Atoi(r.Package.ID)
```

Comment:
won't block the PR over this, but nit: `pkgID` or `rPkgID` (what does the `r` stand for?)

Reply:
> what does the r stand for?
"record", as in "the current claircore record we're working with." I think I was creating the spdx package ID and saving that to `pkgId` in some other commit.